Example of adding custom C functions

guberathome opened the pull request:
One short demonstration how to compile custom code with PForth for the "unix" build.
Custom code resides in separate "custom" folder.
Setup, compilation and cleanup are handled by one shell script.


Tested with MSYS2-Cygwin, Linux, FreeBSD and NetBSD.

Edit: Changes from #184 seem to have crept in here, these still need to be fixed. Apologies.

guberathome committed:
bugfix: removed warnings in WriteCellBigEndian() for 32bit builds

guberathome committed:
Merge branch 'philburk:master' into fix-gub-241230

guberathome committed:
feature: custom demo cf_demo1

guberathome committed:
tested and fixed cf_demo1 on MSYS2, Linux, FreeBSD, NetBSD

guberathome committed:
feature: new custom code handling for unix build

guberathome committed:
Merge branch 'philburk:master' into docu-241230

guberathome committed:
bugfix: latest merge from upstream simplifies FreeBSD handling

guberathome committed:
refactored get-make-cmd to separate shell script

philburk commented:
> Changes from https://github.com/philburk/pforth/pull/184 seem to have crept in here,

there have been several recent changes. You will need to rebase to clear the merge conflicts.

guberathome committed:
merge: manually resolved conflicts with upstream repo

guberathome commented:
> there have been several recent changes. You will need to rebase to clear the merge conflicts.
Done, changes are merged.

philburk commented on custom/cf_helpers.h:
```diff
@@ -0,0 +1,42 @@
+{/* copy PForth string to C-string (zero terminated) 
+    Don't forget to free() the result!
+    TODO: check if there is already defined a similar function in pforth
+ */
```
This is example code.
We should not encourage calling exit() or assert() in production code.
Just return a NULL if the allocation fails.

philburk commented on platforms/unix/Makefile:
```diff
@@ -39,6 +39,13 @@ FULL_WARNINGS =  \
         -Wmissing-declarations
 
+# custom code options
+ifndef CF_SOURCES
```
CF_ is a bit cryptic. We should use CUSTOM_SOURCES

philburk commented on custom/01-be-gone/cf_demo1.c:
```diff
@@ -0,0 +1,99 @@
+ */
+    int res;
+    char* buf = to_C_string( fileName, fnLen );
+    res = remove(buf);  /* delete file in file system */
```
Deleting files is a bit dangerous for a demo.
Better to just do something innocuous.

philburk commented on custom/01-be-gone/go-v1.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/sh
+
+# Compile pForth with custom code and show that this works. 
```
Why v0 and v1 ?

philburk commented on custom/01-be-gone/go-v0.sh:
```diff
@@ -0,0 +1,42 @@
+# save original C sources and copy demo sources. Thus we do not need to change the make file.
+mv ../../csrc/pfcustom.c ../../csrc/pfcustom_c.original
+cp ../cf_helpers.h  ../../csrc/
+cp cf_demo1.c       ../../csrc/pfcustom.c
```
Why overwrite the file?
Can we use the CF_SOURCES (or CUSTOM_SOURCES) in the Makefile?

philburk commented on custom/01-be-gone/demo.fth:
```diff
@@ -0,0 +1,18 @@
+1000 f4711 . ." )"
+CR
+
+\ example of passing passing strings from PForth to custom C code
```
Just need one "passing"

philburk replied:
Thanks for making the other changes.
This is still unresolved. I know that you disagree. But I do not understand why saying "passing passing" is necessary.

philburk reviewed (changes requested):
The "custom" folder should be in an "examples" folder.

guberathome committed:
Merge branch 'philburk:master' into docu-241230

guberathome committed:
from-codereview: moved to examples/custom/ folder

guberathome committed:
from-codereview: replaced BE-GONE by FILE-INFO, fixed go-v1.sh

guberathome committed:
from-codereview: removed panic() and safeAlloc()

guberathome committed:
from-codereview:  renamed Makefile variable to CUSTOM_SOURCES

guberathome committed:
from-codereview: removed go-v0.sh

guberathome commented:
done:
- The "custom" folder should be in an "examples" folder.
- in cf_helpers.h:  We should not encourage calling exit() or assert() in production code.
- in Makefile: We should not encourage calling exit() or assert() in production code.
- in cf_demo1.c: Deleting files is a bit dangerous for a demo 
replaced BE-GONE with FILE-INFO
- go-v1.sh: Why v0 and v1?
- in go-v0.sh: Why overwrite the file?
removed since we changed the Makefile

open:
- in demo.fth: Just need one "passing"
Sorry to disagree: passing more than just one value helps to ensure proper function. 
Also you can guess (by inductive reasoning) what F4711 and FILE-INFO are meant to do just by looking at the output (not the source):
```
---------------------------
show that custom code works
---------------------------
f4711( 0, 1, 10, 100, 1000 ) = ( 11 , 58 , 481 , 4711 , 47011 )
FILE-INFO:
   file{ size=5459, path='Makefile' }
   directory{ path='../../examples' }
   error{ id=2, desc='No such file or directory', path='fileNotHere' }
```

guberathome committed:
fixed typo

guberathome committed:
bugfix: removed superfluos code for MSYS

guberathome commented:
> Just need one "passing"

Please consider if these changes are relevant to be merged (discussion on #188 contains my pitch in favour).
My preference would be to contribute to this repo rather than putting example elsewhere.

I can removed the last offending passages from custom/01-be-gone/demo.fth if that settles it.

philburk commented on examples/custom/01-parameter-passing/cf_demo1.c:
```diff
@@ -0,0 +1,110 @@
+}
+
+static cell_t FileInfo( cell_t path_caddr, cell_t path_len ) {
+	char* path = to_C_string( path_caddr, path_len );
```
This function uses tabs for indentation.
Please use 4 spaces. See:
https://github.com/philburk/pforth/wiki/Style-Guide

philburk commented on examples/custom/01-parameter-passing/demo.fth:
```diff
@@ -0,0 +1,17 @@
+1000 f4711 . ." )"
+CR
+
+\ example of passing passing strings from PForth to custom C code
```
Still "passing passing".

guberathome replied:
Sorry, did not realize the double word.

Done.

philburk commented on examples/custom/cf_helpers.h:
```diff
@@ -0,0 +1,26 @@
+    return buf;
+}
+
+#endif  /* CF_HELPERS_H */
```
Needs EOL

guberathome replied:
Done.

philburk commented on examples/custom/01-parameter-passing/cf_demo1.c:
```diff
@@ -0,0 +1,110 @@
+			asprintf( &result, fmtFile, info.st_size, path );
+	}
+    PUSH_DATA_STACK( (cell_t) result );
+	return (cell_t)strlen(result);
```
path is not getting freed.
Can you avoid doing the dangerous memory allocation for the C string? Maybe allocate on the stack.

guberathome replied:
done, added FREE-C word.

philburk commented on examples/custom/01-parameter-passing/cf_demo1.c:
```diff
@@ -0,0 +1,110 @@
+	char* result;
+	/* MSYS/Cygwin may warn than asprintf() is not defined but compile and run just fine :-/ */
+	if( stat(path, &info) == -1 )
+		asprintf( &result, fmtErr, errno, strerror(errno), path );
```
This does not go through the Forth output path so it will not work over a serial port.

Demonstrating custom C calling does not require calling complex host functions.
It could just be a function that modifies a string.
The user will just replace these custom functions so the simpler the better.

guberathome replied:
> This does not go through the Forth output path so it will not work over a serial port.

Just to clarify: asprintf() allocates a buffer of correct size into which it sprintf's the formatted output.
Which then gets returned. 
I'm not aware of bypassing any PForth functionality here.
 
> Demonstrating custom C calling does not require calling complex host functions. It could just be a function that modifies a string. The user will just replace these custom functions so the simpler the better.

That's what is done here.
Creating a new string (of previously unknown length) requires either hairy assumptions about maximal length or allocating a new buffer. All the fun of programming C.

guberathome committed:
Merge branch 'philburk:master' into docu-241230

guberathome committed:
changes from review

guberathome committed:
free C buffer

guberathome committed:
added description for FileInfo()

guberathome committed:
use result of asprintf instead of calling strlen

guberathome committed:
Merge branch 'master' into docu-241230
